kms/capi/capi.go

```diff
@@ -288,32 +289,205 @@ func (k *CAPIKMS) Close() error {
+			return nil, apiv1.NotFoundError{Message: fmt.Sprintf("certificate with %s=%s not found", KeyIDArg, keyID)}
+		}
+	case issuerName != "" && (serialNumber != "" || subjectCN != ""):
+		var prevCert *windows.CertContext
+		for {
+			handle, err = findCertificateInStore(st,
+				encodingX509ASN|encodingPKCS7,
+				0,
+				findIssuerStr,
+				uintptr(unsafe.Pointer(wide(issuerName))), prevCert)
+
+			if err != nil {
+				return nil, fmt.Errorf("findCertificateInStore failed: %w", err)
+			}
+
+			if handle == nil {
+				return nil, apiv1.NotFoundError{Message: fmt.Sprintf("certificate with %s=%q and %s=%q not found", IssuerNameArg, issuerName, SerialNumberArg, serialNumber)}
+			}
+
+			x509Cert, err := certContextToX509(handle)
+
+			if err != nil {
+				return nil, fmt.Errorf("could not unmarshal certificate to DER: %w", err)
+			}
+
+			switch {
+			case len(serialNumber) > 0:
+				// TODO: Replace this search with a CERT_ID + CERT_ISSUER_SERIAL_NUMBER search instead
+				// https://learn.microsoft.com/en-us/windows/win32/api/wincrypt/ns-wincrypt-cert_id
+				// https://learn.microsoft.com/en-us/windows/win32/api/wincrypt/ns-wincrypt-cert_issuer_serial_number
+				var serialBytes []byte
+				if strings.HasPrefix(serialNumber, "0x") {
+					serialNumber = strings.TrimPrefix(serialNumber, "0x")
+					serialNumber = strings.TrimPrefix(serialNumber, "00") // Comparison fails if leading 00 is not removed
+					serialBytes, err = hex.DecodeString(serialNumber)
+					if err != nil {
+						return nil, fmt.Errorf("invalid hex format for %s: %w", SerialNumberArg, err)
+					}
+				} else {
+					bi := new(big.Int)
+					bi, ok := bi.SetString(serialNumber, 10)
+					if !ok {
+						return nil, fmt.Errorf("invalid %s - must be in hex or integer format", SerialNumberArg)
+					}
+					serialBytes = bi.Bytes()
+				}
+
+				if bytes.Equal(x509Cert.SerialNumber.Bytes(), serialBytes) {
+					return handle, nil
+				}
+			case len(subjectCN) > 0:
+				if x509Cert.Subject.CommonName == subjectCN {
+					return handle, nil
+				}
+			}
+
+			prevCert = handle
+		}
+	default:
+		return nil, fmt.Errorf("%q, %q, or %q and %q is required to find a certificate", HashArg, KeyIDArg, IssuerNameArg, SerialNumberArg)
+	}
+
+	return handle, err
+}
+
 // CreateSigner returns a crypto.Signer that will sign using the key passed in via the URI.
 func (k *CAPIKMS) CreateSigner(req *apiv1.CreateSignerRequest) (crypto.Signer, error) {
 	u, err := uri.ParseWithScheme(Scheme, req.SigningKey)
 	if err != nil {
 		return nil, fmt.Errorf("failed to parse URI: %w", err)
 	}
 
-	var containerName string
-	if containerName = u.Get(ContainerNameArg); containerName == "" {
-		// generate a uuid for the container name
-		containerName, err = randutil.UUIDv4()
+	var (
+		kh            uintptr
+		certHandle    *windows.CertContext
+		containerName string
+	)
+	if containerName = u.Get(ContainerNameArg); containerName != "" {
+		kh, err = nCryptOpenKey(k.providerHandle, containerName, 0, 0)
```

Comment:
Although the error is managed later, it is not quite clear, we can just managed here.

Reply:
1947ee1